boot/header.S

```diff
@@ -21,63 +41,750 @@
+	subw	$MAIN_SEG, %ax		# have we loaded all yet?
+	cmpw	sys_size, %ax
+	jbe	ok1_read
+	ret
+ok1_read:
+	movw	%cs:sectors, %ax
+	subw	sread, %ax
+	movw	%ax, %cx
+	shlw	$9, %cx
+	addw	%bx, %cx
+	jnc	ok2_read
+	je	ok2_read
+	xorw	%ax, %ax
+	subw	%bx, %ax
+	shrw	$9, %ax
+ok2_read:
+	call	read_track
+	movw	%ax, %cx
+	add	sread, %ax
+	cmpw	%cs:sectors, %ax
+	jne	ok3_read
+	movw	$1, %ax
+	subw	head, %ax
+	jne	ok4_read
+	incw	track
+ok4_read:
+	movw	%ax, head
+	xorw	%ax, %ax
+ok3_read:
+	movw	%ax, sread
+	shlw	$9, %cx
+	addw	%cx, %bx
+	jnc	rp_read
+	movw	%es, %ax
+	addb	$0x10, %ah
+	movw	%ax, %es
+	xorw	%bx, %bx
+	jmp	rp_read
+
+read_track:
+	pusha
+	pusha
+	movw	$0xe2e, %ax		# loading... message 2e = .
+	movw	$7, %bx
+	int	$0x10
+	popa
+
+	movw	track, %dx
+	movw	sread, %cx
+	incw	%cx
+	movb	%dl, %ch
+	movw	head, %dx
+	movb	%dl, %dh
+	andw	$0x0100, %dx
+	movb	$2, %ah
+
+	pushw	%dx			# save for error dump
+	pushw	%cx
+	pushw	%bx
+	pushw	%ax
+
+	int	$0x13
+	jc	bad_rt
+	addw	$8, %sp
+	popa
+	ret
+
+bad_rt:
+	pushw	%ax			# save error code
+	call	print_all		# ah = error, al = read
+
+	xorb	%ah, %ah
+	xorb	%dl, %dl
+	int	$0x13
+
+	addw	$10, %sp
+	popa
+	jmp read_track
+
+# This subroutine is for debugging purposes. It will print out all of the
+# registers.  The assumption is that this is called from a routine, with a
+# stack frame like:
+#	dx
+#	cx
+#	bx
+#	ax
+#	err
+#	ret <- sp
+
+print_all:
+	movw	$5, %cx 		# error code + 4 registers
+	movw	%sp, %bp
+
+print_loop:
+	pushw	%cx			# save count left
+	call	print_nl		# nl for readability
+
+	cmpb	5, %cl			# see if register name is needed
+	jae	no_reg
+
+	movw	$(0xe05 + 'A' - 1), %ax
+	subb	%cl, %al
+	int	$0x10
+	movb	$'X', %al
+	int	$0x10
+	movb	$':', %al
+	int	$0x10
+
+no_reg:
+	addw	$2, %bp 		# next register
+	call	print_hex		# print it
+	popw	%cx
+	loop	print_loop
+	ret
+
+print_nl:
+	movw	$0xe0d, %ax		# CR
+	int	$0x10
+	movb	$0x0a, %al		# LF
+	int	$0x10
+	ret
+
+# This subroutine is for debugging purposes, and prints the word pointed to
+# by ss:bp in hexadecimal.
+
+print_hex:
+	movw	$4, %cx 		# 4 hex digits
+	movw	(%bp), %dx		# load word into dx
+
+print_digit:
+	rolw	$4, %dx 		# rotate so that lowest 4 bits are used
+	movb	$0xe, %ah
+	movb	%dl, %al		# mask off so we have only next nibble
+	andb	$0xf, %al
+	addb	$'0', %al		# convert to 0-based digit
+	cmpb	$'9', %al		# check for overflow
+	jbe	good_digit
+	addb	$('A' - '0' - 10), %al
+
+good_digit:
+	int	$0x10
+	loop	print_digit
+	ret
+
+# This subroutine turns off the floppy drive motor, so that we enter the
+# kernel in a known state, and don't have to worry about it later.
+
+kill_motor:
+	pushw	%dx
+	movw	$0x3f2, %dx
+	xorb	%al, %al
+	outb	%al, %dx
+	popw	%dx
+	ret
+
+# This subroutine turns off the text display cursor.
+
+turn_off_cursor:
+	movb	$0x01, %ah
+	movb	$0x00, %bh
+	movw	$0x2000, %cx
+	int	$0x10
+	ret
+
+# Local variables.
+
+sectors:
+	.word	0
+
+
+# Emulate the Linux boot header, to allow loading by intermediate boot loaders.
```

Comment:
Include the couple of extra comment lines from setup.S, suitably modified.

Reply:
Done.